datafusion/functions-aggregate-common/src/aggregate/groups_accumulator.rs

```diff
@@ -367,6 +379,289 @@ impl<T> VecAllocExt for Vec<T> {
+    ///     similar as `take_needed`.
+    fn take_needed_from_blocks<T>(
+        &self,
+        blocks: &mut VecBlocks<T>,
+        mode: GroupStatesMode,
+    ) -> Vec<T>;
+}
+
+impl EmitToExt for EmitTo {
+    fn take_needed<T>(&self, v: &mut Vec<T>) -> Vec<T> {
+        match self {
+            Self::All => {
+                // Take the entire vector, leave new (empty) vector
+                std::mem::take(v)
+            }
+            Self::First(n) => {
+                let split_at = min(v.len(), *n);
+                // get end n+1,.. values into t
+                let mut t = v.split_off(split_at);
+                // leave n+1,.. in v
+                std::mem::swap(v, &mut t);
+                t
+            }
+            Self::NextBlock(_) => unreachable!(
+                "can not support blocked emission in take_needed, you should use take_needed_from_blocks"
+            ),
+        }
+    }
+
+    fn take_needed_from_blocks<T>(
+        &self,
+        blocks: &mut VecBlocks<T>,
+        mode: GroupStatesMode,
+    ) -> Vec<T> {
+        match self {
+            Self::All => {
+                debug_assert!(matches!(mode, GroupStatesMode::Flat));
+                blocks.pop_first_block().unwrap_or_default()
+            }
+            Self::First(n) => {
+                debug_assert!(matches!(mode, GroupStatesMode::Flat));
+
+                let block = blocks.current_mut().unwrap();
+                let split_at = min(block.len(), *n);
+
+                // get end n+1,.. values into t
+                let mut t = block.split_off(split_at);
+                // leave n+1,.. in v
+                std::mem::swap(block, &mut t);
+                t
+            }
+            Self::NextBlock(_) => {
+                debug_assert!(matches!(mode, GroupStatesMode::Blocked(_)));
+                blocks.pop_first_block().unwrap_or_default()
+            }
+        }
+    }
+}
+
+/// Blocked style group index used in blocked mode group values and accumulators
+///
+/// Parts in index:
+///   - High 32 bits represent `block_id`
+///   - Low 32 bits represent `block_offset`
+#[derive(Debug, Clone, Copy)]
+pub struct BlockedGroupIndex {
+    pub block_id: usize,
+    pub block_offset: usize,
+}
+
+impl BlockedGroupIndex {
+    pub fn new(group_index: usize) -> Self {
+        let block_id =
+            ((group_index as u64 >> 32) & BLOCKED_INDEX_LOW_32_BITS_MASK) as usize;
+        let block_offset =
+            ((group_index as u64) & BLOCKED_INDEX_LOW_32_BITS_MASK) as usize;
+
+        Self {
+            block_id,
+            block_offset,
+        }
+    }
+
+    pub fn new_from_parts(block_id: usize, block_offset: usize) -> Self {
+        Self {
+            block_id,
+            block_offset,
+        }
+    }
+
+    pub fn as_packed_index(&self) -> usize {
+        ((((self.block_id as u64) << 32) & BLOCKED_INDEX_HIGH_32_BITS_MASK)
```

Comment:
The masking doesn't seem necessary?
Why not just `(self.block_id as u64) << 32) | (self.block_offset as u64)` ?

Reply:
Yes, `(self.block_id as u64) << 32) | (self.block_offset as u64)` is enough for me, too.
But I see in bit mask of wanted bits will ususally used after bit operations in c codes, I am not sure about should we perform like this.

Reply:
I think it's a no-op in this case, so seems better to remove it.

Reply:
fixed